### PR TITLE
feat: create variant composable

### PR DESCRIPTION
**Pull request by johnleider**
## Description
Add new variant composable with 4 options:

* **plain** - low opacity with no background or elevation. raises opacity on hover
* **contained** - background with elevation
* **outlined** - low opacity colored border with no elevation
* **text** - uses current text color with lowered opacity for background

Existing usage of outlined for border has been removed. **variant="outlined"** differs from **border** in that it has a default border opacity of *1*, whereas **border** starts at the default divider opacity, **0.12**.

## Motivation and Context
Reduce prop overhead by normalizing the some of the different styles within the framework.

## How Has This Been Tested?

unit

## Markup:

<details>

```vue
<template>
  <v-app>
    <v-container>
      <div class="text-h5 ma-2">v-btn</div>
      <template v-for="variant in ['contained', 'outlined', 'text', 'plain']" :key="variant">
        <div class="mb-12">
          <v-btn class="mr-6" :variant="variant" link>
            {{ variant }} variant
          </v-btn>

          <v-btn color="primary" :variant="variant" link>
            {{ variant }} variant
          </v-btn>
        </div>
      </template>

      <div class="text-h5 ma-2">v-list</div>
      <template v-for="variant in ['contained', 'outlined', 'text', 'plain']" :key="variant">
        <v-list :subheader="variant + ' variant'">
          <v-list-item title="Default" link :variant="variant" />
          <v-list-item title="Colored" link :variant="variant" color="primary" />
        </v-list>
      </template>

      <div class="text-h5 ma-2">v-card</div>
      <template v-for="variant in ['contained', 'outlined', 'text', 'plain']" :key="variant">
        <div class="mb-12">
          <v-card class="mb-6" :variant="variant" link>
            <v-card-header>
              <v-card-header-text>
                <v-card-title>{{ variant }} variant</v-card-title>

                <v-card-subtitle>Default</v-card-subtitle>
              </v-card-header-text>
            </v-card-header>

            <v-card-text>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Quaerat aperiam hic quo aspernatur nihil quidem corporis asperiores numquam repudiandae tempore, ad voluptates non deleniti.
            </v-card-text>
          </v-card>

          <v-card color="primary" :variant="variant" link>
            <v-card-header>
              <v-card-header-text>
                <v-card-title>{{ variant }} variant</v-card-title>

                <v-card-subtitle>Colored</v-card-subtitle>
              </v-card-header-text>
            </v-card-header>

            <v-card-text>
              Lorem ipsum dolor sit amet consectetur adipisicing elit. Quaerat aperiam hic quo aspernatur nihil quidem corporis asperiores numquam repudiandae tempore, ad voluptates non deleniti.
            </v-card-text>
          </v-card>
        </div>
      </template>
    </v-container>
  </v-app>
</template>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/composables/variant.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  variant: {
+    type: String as PropType<Variant>,
+    default: 'default',
+    validator: (v: any) => allowedVariants.includes(v),
+  },
+}, 'variant')
+
+export function useVariant (props: VariantProps, name: string) {
+  const variantClasses = computed(() => {
+    return `${name}--variant-${props.variant}`
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/composables/variant.ts:31]:
```suggestion
    return props.variant ? `${name}--variant-${props.variant}` : ''
```

Comment (johnleider) [reply packages/vuetify/src/composables/variant.ts:31]: Any component that uses variants should have a default variant class. Same as density

File: packages/vuetify/src/components/VCard/VCard.tsx [hunk 2]
```diff
@@ -62,17 +64,18 @@ export default defineComponent({
     ...makePositionProps(),
     ...makeRoundedProps(),
     ...makeTagProps(),
+    ...makeVariantProps({ variant: 'contained' as Variant }),
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCard/VCard.tsx:67]:
```suggestion
    ...makeVariantProps({ variant: 'contained' } as const),
```

File: packages/vuetify/src/components/VBtn/VBtn.tsx [hunk 3]
```diff
@@ -51,37 +52,30 @@ export default defineComponent({
     ...makeSizeProps(),
     ...makeTagProps({ tag: 'button' }),
     ...makeThemeProps(),
+    ...makeVariantProps({ variant: 'contained' as Variant }),
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VBtn/VBtn.tsx:55]:
```suggestion
    ...makeVariantProps({ variant: 'contained' } as const),
```

File: packages/vuetify/src/components/VBtn/VBtn.tsx [hunk 4]
```diff
@@ -51,37 +49,30 @@ export default defineComponent({
...
+    const { colorClasses, colorStyles, variantClasses } = useVariant(props, 'v-btn')
     const { roundedClasses } = useRounded(props, 'v-btn')
     const { densityClasses } = useDensity(props, 'v-btn')
     const { dimensionStyles } = useDimension(props)
     const { elevationClasses } = useElevation(props)
     const { positionClasses, positionStyles } = usePosition(props, 'v-btn')
     const { sizeClasses } = useSize(props, 'v-btn')
 
-    const isContained = computed(() => {
-      return !(props.text || props.plain || props.outlined || props.border !== false)
```
Comment (KaelWD) [inline packages/vuetify/src/components/VBtn/VBtn.tsx:67]: We still have text/flat/plain/outlined props?

File: packages/vuetify/src/composables/variant.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  })
+
+  const { colorClasses, colorStyles } = useColor(computed(() => ({
+    [props.variant === 'contained' ? 'background' : 'text']: props.color,
+  })))
+
+  return { colorClasses, colorStyles, variantClasses }
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/composables/variant.ts:38]: is there a scenario where someone would use only `colorClasses` but not `variantClasses` or vice versa?

Comment (johnleider) [reply packages/vuetify/src/composables/variant.ts:38]: Not that I can think of. I wanted to be verbose with it.

File: packages/vuetify/src/components/VBtn/VBtn.tsx (hunk 4 above)
Comment (johnleider) [reply packages/vuetify/src/components/VBtn/VBtn.tsx:67]: Stylistically nothing has changed. Instead of being handled by javascript it's done by css now.

**Comment by KaelWD** [comment]
> text - lowered opacity background with no elevation

This doesn't really make sense? `<v-btn text>` and `<v-list-item>` both just have text and a hover overlay, with a transparent background.

**Comment by johnleider** [comment]
> > text - lowered opacity background with no elevation
> 
> This doesn't really make sense? `<v-btn text>` and `<v-list-item>` both just have text and a hover overlay, with a transparent background.

Background color is determined based upon the text color when using variant text. I'll update the language.

**Comment by johnleider** [comment]
Requesting **Final** comments from @vuetifyjs/core-team

**Pull request merged**